File: src/data/types/purchaseCapacityResult.ts
import type { CrewCapacity } from "./crewCapacity.ts";
import type { Wallet } from "./wallet.ts";

// Necessary completion, discovered while implementing Agent 18: its
// contract requires a UI "option to purchase additional capacity" (Phase
// 4 GDD §2.4), but Agent 16's contract never listed a corresponding
// function -- §2.4 decided the mechanic, but no Outputs entry for it
// exists in agent-16-crew-core.md. Since Agent 18 must never implement
// crew logic itself, this belongs in Agent 16's module (src/crew/), not
// presentation -- see purchaseCapacity.ts.
export interface PurchaseCapacitySucceeded {
  purchased: true;
  updatedCapacity: CrewCapacity;
  updatedWallet: Wallet;
}

export interface PurchaseCapacityRejected {
  purchased: false;
  reason: string;
}

export type PurchaseCapacityResult = PurchaseCapacitySucceeded | PurchaseCapacityRejected;

File: unity/ProfitableCore/Schema/PurchaseCapacityResult.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/purchaseCapacityResult.ts's PurchaseCapacityResult
// = PurchaseCapacitySucceeded | PurchaseCapacityRejected discriminated
// union.
public abstract class PurchaseCapacityResult
{
    public abstract bool Purchased { get; }
}

public sealed class PurchaseCapacitySucceeded : PurchaseCapacityResult
{
    public override bool Purchased => true;
    public CrewCapacity UpdatedCapacity { get; init; } = new();
    public Wallet UpdatedWallet { get; init; } = new();
}

public sealed class PurchaseCapacityRejected : PurchaseCapacityResult
{
    public override bool Purchased => false;
    public string Reason { get; init; } = string.Empty;
}
